:farmer: Kci__nightly-release/debug_ubuntu_noble `sros2.test.sros2.commands.security` test failing in zenoh caused by an assertion

### Generated by Generative AI

_No response_

### Operating System:

Ubuntu 24.04

### ROS version or commit hash:

Kilted

### RMW implementation (if applicable):

rmw_zenoh_cpp

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

rclcpp

### 'ros2 doctor --report' output

_No response_

### Steps to reproduce issue

1. Go to http://build.ros2.org/job/Kci__nightly-debug_ubuntu_noble_amd64 or http://build.ros2.org/job/Kci__nightly-release_ubuntu_noble_amd64
2. Build a job
3. See test failing

### Expected behavior

To buiild without test regressions

### Actual behavior

`sros2.test.sros2.commands.security.verbs.test_generate_policy.test_generate_policy` test failing

### Additional information

### Reference Build
- [Kci__nightly-debug_ubuntu_noble_amd64#77](http://build.ros2.org/job/Kci__nightly-debug_ubuntu_noble_amd64/77)
- [Kci__nightly-release_ubuntu_noble_amd64#76](http://build.ros2.org/job/Kci__nightly-release_ubuntu_noble_amd64/76)

### Failing tests
- [sros2.test.sros2.commands.security.verbs.test_generate_policy.test_generate_policy](https://build.ros2.org/job/Kci__nightly-release_ubuntu_noble_amd64/71/testReport/junit/sros2.test.sros2.commands.security.verbs/test_generate_policy/test_generate_policy/)

### log Output
```
test_generate_policy.TestSROS2GeneratePolicyVerb.test_generate_policy_services[rmw_zenoh_cpp, True] failed
test_generate_policy.TestSROS2GeneratePolicyVerb.test_generate_policy_topics[rmw_zenoh_cpp, True] failed
test_generate_policy.TestSROS2GeneratePolicyVerb.test_generate_policy_services[rmw_zenoh_cpp, False] failed
test_generate_policy.TestSROS2GeneratePolicyVerb.test_generate_policy_topics[rmw_zenoh_cpp, False] failed

==================================================================================================================================================================================================================================================================================================================================================================================================
FAIL: test_generate_policy.TestSROS2GeneratePolicyVerb.test_generate_policy_services[rmw_zenoh_cpp, True]
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/ws/src/ros2/sros2/sros2/test/sros2/commands/security/verbs/test_generate_policy.py", line 159, in test_generate_policy_services
    assert self.wait_for(
AssertionError: assert False
 +  where False = <bound method SROS2CLITestCase.setUpClass.<locals>.wait_for of <test_generate_policy.TestSROS2GeneratePolicyVerb testMethod=test_generate_policy_services[rmw_zenoh_cpp, True]>>(expected_nodes=['/node_ns/test_generate_policy_services_node'], expected_services=['/node_ns/test_generate_policy_services_node/client', '/node_ns/test_generate_policy_services_node/server'])
 +    where <bound method SROS2CLITestCase.setUpClass.<locals>.wait_for of <test_generate_policy.TestSROS2GeneratePolicyVerb testMethod=test_generate_policy_services[rmw_zenoh_cpp, True]>> = <test_generate_policy.TestSROS2GeneratePolicyVerb testMethod=test_generate_policy_services[rmw_zenoh_cpp, True]>.wait_for


==================================================================================================================================================================================================================================================================================================================================================================================================
FAIL: test_generate_policy.TestSROS2GeneratePolicyVerb.test_generate_policy_topics[rmw_zenoh_cpp, True]
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/ws/src/ros2/sros2/sros2/test/sros2/commands/security/verbs/test_generate_policy.py", line 112, in test_generate_policy_topics
    assert self.wait_for(
AssertionError: assert False
 +  where False = <bound method SROS2CLITestCase.setUpClass.<locals>.wait_for of <test_generate_policy.TestSROS2GeneratePolicyVerb testMethod=test_generate_policy_topics[rmw_zenoh_cpp, True]>>(expected_nodes=['//test_generate_policy_topics_node'], expected_topics=['/test_generate_policy_topics_node/pub', '/test_generate_policy_topics_node/sub'])
 +    where <bound method SROS2CLITestCase.setUpClass.<locals>.wait_for of <test_generate_policy.TestSROS2GeneratePolicyVerb testMethod=test_generate_policy_topics[rmw_zenoh_cpp, True]>> = <test_generate_policy.TestSROS2GeneratePolicyVerb testMethod=test_generate_policy_topics[rmw_zenoh_cpp, True]>.wait_for


==================================================================================================================================================================================================================================================================================================================================================================================================
FAIL: test_generate_policy.TestSROS2GeneratePolicyVerb.test_generate_policy_services[rmw_zenoh_cpp, False]
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/ws/src/ros2/sros2/sros2/test/sros2/commands/security/verbs/test_generate_policy.py", line 173, in test_generate_policy_services
    assert gen_command.exit_code == launch_testing.asserts.EXIT_OK
AssertionError: assert 1 == 0
 +  where 1 = <launch_testing.tools.process.ProcessProxy object at 0x7511456696a0>.exit_code
 +  and   0 = <module 'launch_testing.asserts' from '/tmp/ws/install_isolated/launch_testing/lib/python3.12/site-packages/launch_testing/asserts/__init__.py'>.EXIT_OK
 +    where <module 'launch_testing.asserts' from '/tmp/ws/install_isolated/launch_testing/lib/python3.12/site-packages/launch_testing/asserts/__init__.py'> = launch_testing.asserts

==================================================================================================================================================================================================================================================================================================================================================================================================
FAIL: test_generate_policy.TestSROS2GeneratePolicyVerb.test_generate_policy_topics[rmw_zenoh_cpp, False]
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/ws/src/ros2/sros2/sros2/test/sros2/commands/security/verbs/test_generate_policy.py", line 126, in test_generate_policy_topics
    assert gen_command.exit_code == launch_testing.asserts.EXIT_OK
AssertionError: assert 1 == 0
 +  where 1 = <launch_testing.tools.process.ProcessProxy object at 0x75114566a6f0>.exit_code
 +  and   0 = <module 'launch_testing.asserts' from '/tmp/ws/install_isolated/launch_testing/lib/python3.12/site-packages/launch_testing/asserts/__init__.py'>.EXIT_OK
 +    where <module 'launch_testing.asserts' from '/tmp/ws/install_isolated/launch_testing/lib/python3.12/site-packages/launch_testing/asserts/__init__.py'> = launch_testing.asserts
```

### Flakyness last 15 days
job_name|last_fail|first_fail|build_count|failure_count|failure_percentage
|--|--|--|--|--|--|
Kci__nightly-release_ubuntu_noble_amd64|2025-12-19|2025-12-07|5|5|100.0
Kci__nightly-debug_ubuntu_noble_amd64|2025-12-19|2025-12-07|5|5|100.0

CC:  @Crola1702

--- a/sros2/test/sros2/commands/security/verbs/utilities/sros2_cli_test_case.py
+++ b/sros2/test/sros2/commands/security/verbs/utilities/sros2_cli_test_case.py
@@ -15,8 +15,11 @@
 import argparse
 import contextlib
 import time
+from typing import cast
 import unittest
 
+
+from launch import Action
 from launch import LaunchDescription
 from launch.actions import ExecuteProcess
 from launch.actions import RegisterEventHandler
@@ -44,7 +47,7 @@ def generate_sros2_cli_test_description(
     additional_env = get_rmw_additional_env(rmw_implementation)
     set_env_actions = [SetEnvironmentVariable(k, v) for k, v in additional_env.items()]
     # Build shutdown actions based on whether daemon is used
-    shutdown_actions = []
+    shutdown_actions: list[Action] = []
     if use_daemon:
         # Stop daemon in isolated environment with proper ROS_DOMAIN_ID
         shutdown_actions = [
@@ -52,7 +55,7 @@ def generate_sros2_cli_test_description(
                 cmd=['ros2', 'daemon', 'stop'],
                 name='daemon-stop-isolated',
                 # Use the same isolated environment
-                additional_env=dict(additional_env),
+                additional_env=cast(dict, additional_env),
             ),
         ]
     shutdown_actions.append(ResetEnvironment())

--- a/sros2/test/sros2/commands/security/verbs/utilities/sros2_cli_test_case.py
+++ b/sros2/test/sros2/commands/security/verbs/utilities/sros2_cli_test_case.py
@@ -43,6 +43,19 @@ def generate_sros2_cli_test_description(
 ) -> LaunchDescription:
     additional_env = get_rmw_additional_env(rmw_implementation)
     set_env_actions = [SetEnvironmentVariable(k, v) for k, v in additional_env.items()]
+    # Build shutdown actions based on whether daemon is used
+    shutdown_actions = []
+    if use_daemon:
+        # Stop daemon in isolated environment with proper ROS_DOMAIN_ID
+        shutdown_actions = [
+            ExecuteProcess(
+                cmd=['ros2', 'daemon', 'stop'],
+                name='daemon-stop-isolated',
+                # Use the same isolated environment
+                additional_env=dict(additional_env),
+            ),
+        ]
+    shutdown_actions.append(ResetEnvironment())
     if use_daemon:
         # Start daemon.
         fixture_actions = [ExecuteProcess(
@@ -53,17 +66,7 @@ def generate_sros2_cli_test_description(
     fixture_actions = [
         *set_env_actions,
         EnableRmwIsolation(),
-        RegisterEventHandler(OnShutdown(on_shutdown=[
-            # Stop daemon in isolated environment with proper ROS_DOMAIN_ID
-            ExecuteProcess(
-                cmd=['ros2', 'daemon', 'stop'],
-                name='daemon-stop-isolated',
-                # Use the same isolated environment
-                additional_env=dict(additional_env),
-            ),
-            # This must be done after stopping the daemon in the isolated environment
-            ResetEnvironment(),
-        ])),
+        RegisterEventHandler(OnShutdown(on_shutdown=shutdown_actions)),
         *fixture_actions,
     ]
     return LaunchDescription([

--- a/sros2/test/sros2/commands/security/verbs/utilities/sros2_cli_test_case.py
+++ b/sros2/test/sros2/commands/security/verbs/utilities/sros2_cli_test_case.py
@@ -19,7 +19,10 @@ import unittest
 
 from launch import LaunchDescription
 from launch.actions import ExecuteProcess
+from launch.actions import RegisterEventHandler
+from launch.actions import ResetEnvironment
 from launch.actions import SetEnvironmentVariable
+from launch.event_handlers import OnShutdown
 
 import launch_testing
 import launch_testing.asserts
@@ -50,6 +53,17 @@ def generate_sros2_cli_test_description(
     fixture_actions = [
         *set_env_actions,
         EnableRmwIsolation(),
+        RegisterEventHandler(OnShutdown(on_shutdown=[
+            # Stop daemon in isolated environment with proper ROS_DOMAIN_ID
+            ExecuteProcess(
+                cmd=['ros2', 'daemon', 'stop'],
+                name='daemon-stop-isolated',
+                # Use the same isolated environment
+                additional_env=dict(additional_env),
+            ),
+            # This must be done after stopping the daemon in the isolated environment
+            ResetEnvironment(),
+        ])),
         *fixture_actions,
     ]
     return LaunchDescription([

